src/tensor_ops/utilities/cpu_kernels.rs

```diff
@@ -74,22 +74,127 @@ impl<E: Dtype, Op: BinaryDerivative<E>> BinaryKernel<Op, E> for Cpu {
         grad_rhs: &mut Self::Storage<S, E>,
         grad_out: &Self::Storage<S, E>,
     ) -> Result<(), Self::Err> {
-        let mut lhs_idx = NdIndex::new(lhs.shape, lhs.strides);
-        let mut rhs_idx = NdIndex::new(rhs.shape, rhs.strides);
         let lhs_buf = lhs.data.as_ref();
         let rhs_buf = rhs.data.as_ref();
         let grad_lhs_buf = std::sync::Arc::make_mut(&mut grad_lhs.data);
         let grad_rhs_buf = std::sync::Arc::make_mut(&mut grad_rhs.data);
-        // NOTE: we can use .buf_iter() here because we know the outcome of this op is
-        // contiguous from forward
-        for &go in grad_out.buf_iter() {
-            let lhs_i = lhs_idx.next().unwrap();
-            let rhs_i = rhs_idx.next().unwrap();
-            let l = &lhs_buf[lhs_i];
-            let r = &rhs_buf[rhs_i];
-            grad_lhs_buf[lhs_i] += op.dfdx(l, r) * go;
-            grad_rhs_buf[rhs_i] += op.dfdy(l, r) * go;
+        let out_buf = grad_out.data.as_ref();
+        let numel = lhs.shape.num_elements();
+
+        if lhs.strides == rhs.strides && lhs.strides == grad_out.strides {
+            // contiguous case
+            for i in 0..numel {
+                let l = &lhs_buf[i];
+                let r = &rhs_buf[i];
+                let go = out_buf[i];
+                grad_lhs_buf[i] += op.dfdx(l, r) * go;
+                grad_rhs_buf[i] += op.dfdy(l, r) * go;
+            }
+        } else {
+            let lhs_num_br = numel / lhs.data.len();
+            let rhs_num_br = numel / rhs.data.len();
+
+            if lhs_num_br > rhs_num_br {
+                // lhs has less data - iterate lhs in logical order
+                let [_, mut rhs_idx, mut out_idx] =
+                    index_for_binop(lhs.shape, lhs.strides, rhs.strides);
+                for (l, gl) in lhs_buf.iter().zip(grad_lhs_buf.iter_mut()) {
+                    let mut tmp = Default::default();
+                    for _ in 0..lhs_num_br {
+                        let i_rhs = rhs_idx.next().unwrap();
+                        let i_out = out_idx.next().unwrap();
+                        let r = &rhs_buf[i_rhs];
+                        let go = out_buf[i_out];
+                        tmp += op.dfdx(l, r) * go;
+                        grad_rhs_buf[i_rhs] += op.dfdy(l, r) * go;
+                    }
+                    *gl += tmp;
+                }
+            } else {
+                // rhs has less data - iterate rhs in logical order
+                let [_, mut lhs_idx, mut out_idx] =
+                    index_for_binop(rhs.shape, rhs.strides, lhs.strides);
```

Comment:
For both branches of the if, it's also possible that we can iterate in logical order if lhs_idx/out_idx are contiguous. This happens when the dims that are broadcast are already rightmost, e.g. softmax